File: AppCheckCore/Tests/Unit/DebugProvider/GACAppCheckDebugProviderAPIServiceTests.m
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import <XCTest/XCTest.h>

#import "FBLPromise+Testing.h"

#import "AppCheckCore/Sources/Public/AppCheckCore/GACAppCheck.h"
#import "AppCheckCore/Sources/Public/AppCheckCore/GACAppCheckToken.h"

#import "AppCheckCore/Sources/DebugProvider/API/GACAppCheckDebugProviderAPIService.h"
#import "AppCheckCore/Sources/Public/AppCheckCore/_GACAppCheckAPIService.h"
#import "AppCheckCore/Sources/Public/AppCheckCore/_GACAppCheckErrorUtil.h"
#import "AppCheckCore/Sources/Public/AppCheckCore/_GACURLSessionDataResponse.h"

#import "AppCheckCore/Tests/Unit/Utils/GACAppCheckAPIServiceFake.h"
#import "AppCheckCore/Tests/Unit/Utils/GACURLSessionFake.h"

static NSString *const kResourceName = @"projects/test_project_id/apps/test_app_id";

@interface GACAppCheckDebugProviderAPIServiceTests : XCTestCase
@property(nonatomic) GACAppCheckDebugProviderAPIService *debugAPIService;

@property(nonatomic) GACAppCheckAPIServiceFake *mockAPIService;
@end

@implementation GACAppCheckDebugProviderAPIServiceTests

- (void)setUp {
  [super setUp];

  self.mockAPIService = [[GACAppCheckAPIServiceFake alloc] init];
  self.mockAPIService.baseURL = @"https://test.appcheck.url.com/alpha";

  self.debugAPIService =
      [[GACAppCheckDebugProviderAPIService alloc] initWithAPIService:self.mockAPIService
                                                        resourceName:kResourceName];
}

- (void)tearDown {
  self.debugAPIService = nil;
  self.mockAPIService = nil;
  [super tearDown];
}

- (void)testAppCheckTokenSuccess {
  [self testAppCheckTokenSuccessWithLimitedUse:NO];
}

- (void)testAppCheckTokenSuccessWithLimitedUse {
  [self testAppCheckTokenSuccessWithLimitedUse:YES];
}

- (void)testAppCheckTokenSuccessWithLimitedUse:(BOOL)limitedUse {
  NSString *debugToken = [NSUUID UUID].UUIDString;
  GACAppCheckToken *expectedResult = [[GACAppCheckToken alloc] initWithToken:@"app_check_token"
                                                              expirationDate:[NSDate date]];

  // 1. Stub API service.
  // 1.1. Stub API response.
  NSString *expectedRequestURL =
      [NSString stringWithFormat:@"%@%@", [self.mockAPIService baseURL],
                                 @"/projects/test_project_id/apps/test_app_id:exchangeDebugToken"];
  NSData *fakeResponseData = [@"fake response" dataUsingEncoding:NSUTF8StringEncoding];
  NSHTTPURLResponse *HTTPResponse = [GACURLSessionFake HTTPResponseWithCode:200];
  _GACURLSessionDataResponse *APIResponse =
      [[_GACURLSessionDataResponse alloc] initWithResponse:HTTPResponse HTTPBody:fakeResponseData];

  self.mockAPIService.sendRequestPromise = [FBLPromise resolvedWith:APIResponse];

  // 1.2. Stub response parsing.
  self.mockAPIService.appCheckTokenPromise = [FBLPromise resolvedWith:expectedResult];

  // 2. Send request.
  __auto_type tokenPromise = [self.debugAPIService appCheckTokenWithDebugToken:debugToken
                                                                    limitedUse:limitedUse];

  // 3. Verify.
  XCTAssert(FBLWaitForPromisesWithTimeout(1));

  XCTAssertTrue(tokenPromise.isFulfilled);
  XCTAssertNil(tokenPromise.error);

  XCTAssertEqualObjects(tokenPromise.value.token, expectedResult.token);
  XCTAssertEqualObjects(tokenPromise.value.expirationDate, expectedResult.expirationDate);

  XCTAssertEqualObjects(tokenPromise.value.token, expectedResult.token);
  XCTAssertEqualObjects(tokenPromise.value.expirationDate, expectedResult.expirationDate);

  XCTAssertEqualObjects(self.mockAPIService.passedRequestURL.absoluteString, expectedRequestURL);
  XCTAssertEqualObjects(self.mockAPIService.passedHTTPMethod, @"POST");
  XCTAssertEqualObjects(self.mockAPIService.passedAdditionalHeaders[@"Content-Type"],
                        @"application/json");
  [self assertHTTPBody:self.mockAPIService.passedBody debugToken:debugToken limitedUse:limitedUse];
  XCTAssertEqualObjects(self.mockAPIService.passedAPIResponse, APIResponse);
}

- (void)testAppCheckTokenResponseParsingError {
  NSString *debugToken = [NSUUID UUID].UUIDString;
  NSError *parsingError = [NSError errorWithDomain:@"testAppCheckTokenResponseParsingError"
                                              code:-1
                                          userInfo:nil];

  // 1. Stub API service.
  // 1.1. Stub API response.
  NSString *expectedRequestURL =
      [NSString stringWithFormat:@"%@%@", [self.mockAPIService baseURL],
                                 @"/projects/test_project_id/apps/test_app_id:exchangeDebugToken"];
  NSData *fakeResponseData = [@"fake response" dataUsingEncoding:NSUTF8StringEncoding];
  NSHTTPURLResponse *HTTPResponse = [GACURLSessionFake HTTPResponseWithCode:200];
  _GACURLSessionDataResponse *APIResponse =
      [[_GACURLSessionDataResponse alloc] initWithResponse:HTTPResponse HTTPBody:fakeResponseData];

  self.mockAPIService.sendRequestPromise = [FBLPromise resolvedWith:APIResponse];

  // 1.2. Stub response parsing.
  FBLPromise *rejectedPromise = [FBLPromise pendingPromise];
  [rejectedPromise reject:parsingError];
  self.mockAPIService.appCheckTokenPromise = rejectedPromise;

  // 2. Send request.
  __auto_type tokenPromise = [self.debugAPIService appCheckTokenWithDebugToken:debugToken
                                                                    limitedUse:NO];

  // 3. Verify.
  XCTAssert(FBLWaitForPromisesWithTimeout(1));

  XCTAssertTrue(tokenPromise.isRejected);
  XCTAssertEqualObjects(tokenPromise.error, parsingError);
  XCTAssertNil(tokenPromise.value);

  XCTAssertEqualObjects(self.mockAPIService.passedRequestURL.absoluteString, expectedRequestURL);
  XCTAssertEqualObjects(self.mockAPIService.passedHTTPMethod, @"POST");
  XCTAssertEqualObjects(self.mockAPIService.passedAdditionalHeaders[@"Content-Type"],
                        @"application/json");
  [self assertHTTPBody:self.mockAPIService.passedBody debugToken:debugToken limitedUse:NO];
  XCTAssertEqualObjects(self.mockAPIService.passedAPIResponse, APIResponse);
}

- (void)testAppCheckTokenNetworkError {
  NSString *debugToken = [NSUUID UUID].UUIDString;
  NSError *APIError = [NSError errorWithDomain:@"testAppCheckTokenNetworkError"
                                          code:-1
                                      userInfo:nil];

  // 1. Stub API service.
  FBLPromise *rejectedPromise = [FBLPromise pendingPromise];
  [rejectedPromise reject:APIError];

  self.mockAPIService.sendRequestPromise = rejectedPromise;

  // 2. Send request.
  __auto_type tokenPromise = [self.debugAPIService appCheckTokenWithDebugToken:debugToken
                                                                    limitedUse:NO];

  // 3. Verify.
  XCTAssert(FBLWaitForPromisesWithTimeout(1));

  XCTAssertTrue(tokenPromise.isRejected);
  XCTAssertNil(tokenPromise.value);
  XCTAssertEqualObjects(tokenPromise.error, APIError);

  [self assertHTTPBody:self.mockAPIService.passedBody debugToken:debugToken limitedUse:NO];
}

#pragma mark - Helpores

- (void)assertHTTPBody:(NSData *)body
            debugToken:(NSString *)debugToken
            limitedUse:(BOOL)limitedUse {
  NSDictionary<NSString *, id> *decodedData = [NSJSONSerialization JSONObjectWithData:body
                                                                              options:0
                                                                                error:nil];
  XCTAssert([decodedData isKindOfClass:[NSDictionary class]]);

  NSString *decodeDebugToken = decodedData[@"debug_token"];
  XCTAssertNotNil(decodeDebugToken);
  XCTAssertEqualObjects(decodeDebugToken, debugToken);
  NSNumber *decodedLimitedUse = decodedData[@"limited_use"];
  XCTAssertNotNil(decodedLimitedUse);
  XCTAssertEqualObjects(decodedLimitedUse, @(limitedUse));
}

@end

File: AppCheckCore/Tests/Unit/DebugProvider/GACAppCheckDebugProviderAPIServiceTests.swift
import XCTest
#if canImport(Promises)
import Promises
#endif
@testable import AppCheckCore

class MockAppCheckAPIService: NSObject, _GACAppCheckAPIServiceProtocol {
    var baseURL: String = "https://test.appcheck.url.com/alpha"
    
    var passedRequestURL: URL?
    var passedHTTPMethod: String?
    var passedBody: Data?
    var passedAdditionalHeaders: [String: String]?
    
    var sendRequestResult: Result<_GACURLSessionDataResponse, Error>?
    var appCheckTokenResult: Result<GACAppCheckToken, Error>?
    
    var passedAPIResponse: _GACURLSessionDataResponse?
    
    func sendRequest(with requestURL: URL, httpMethod: String, body: Data?, additionalHeaders: [String : String]?) -> FBLPromise<_GACURLSessionDataResponse> {
        passedRequestURL = requestURL
        passedHTTPMethod = httpMethod
        passedBody = body
        passedAdditionalHeaders = additionalHeaders
        
        let promise = FBLPromise<_GACURLSessionDataResponse>.pending()
        if let result = sendRequestResult {
            switch result {
            case .success(let response): promise.fulfill(response)
            case .failure(let error): promise.reject(error)
            }
        }
        return promise
    }
    
    func appCheckToken(withAPIResponse response: _GACURLSessionDataResponse) -> FBLPromise<GACAppCheckToken> {
        passedAPIResponse = response
        let promise = FBLPromise<GACAppCheckToken>.pending()
        if let result = appCheckTokenResult {
            switch result {
            case .success(let token): promise.fulfill(token)
            case .failure(let error): promise.reject(error)
            }
        }
        return promise
    }
}

class GACAppCheckDebugProviderAPIServiceTests: XCTestCase {
    var debugAPIService: GACAppCheckDebugProviderAPIService!
    var mockAPIService: MockAppCheckAPIService!
    
    let kResourceName = "projects/test_project_id/apps/test_app_id"
    
    override func setUp() {
        super.setUp()
        mockAPIService = MockAppCheckAPIService()
        debugAPIService = GACAppCheckDebugProviderAPIService(apiService: mockAPIService, resourceName: kResourceName)
    }
    
    override func tearDown() {
        debugAPIService = nil
        mockAPIService = nil
        super.tearDown()
    }
    
    func testAppCheckTokenSuccess() async throws {
        try await testAppCheckTokenSuccess(withLimitedUse: false)
    }
    
    func testAppCheckTokenSuccessWithLimitedUse() async throws {
        try await testAppCheckTokenSuccess(withLimitedUse: true)
    }
    
    func testAppCheckTokenSuccess(withLimitedUse limitedUse: Bool) async throws {
        let debugToken = UUID().uuidString
        let expectedResult = GACAppCheckToken(token: "app_check_token", expirationDate: Date())
        
        let expectedRequestURL = "\(mockAPIService.baseURL)/projects/test_project_id/apps/test_app_id:exchangeDebugToken"
        let fakeResponseData = "fake response".data(using: .utf8)!
        let httpResponse = HTTPURLResponse(url: URL(string: expectedRequestURL)!, statusCode: 200, httpVersion: nil, headerFields: nil)!
        let apiResponse = _GACURLSessionDataResponse(response: httpResponse, httpBody: fakeResponseData)
        
        mockAPIService.sendRequestResult = .success(apiResponse)
        mockAPIService.appCheckTokenResult = .success(expectedResult)
        
        let token = try await debugAPIService.appCheckToken(debugToken: debugToken, limitedUse: limitedUse)
        
        XCTAssertEqual(token.token, expectedResult.token)
        XCTAssertEqual(token.expirationDate, expectedResult.expirationDate)
        
        XCTAssertEqual(mockAPIService.passedRequestURL?.absoluteString, expectedRequestURL)
        XCTAssertEqual(mockAPIService.passedHTTPMethod, "POST")
        XCTAssertEqual(mockAPIService.passedAdditionalHeaders?["Content-Type"], "application/json")
        try assertHTTPBody(mockAPIService.passedBody, debugToken: debugToken, limitedUse: limitedUse)
        XCTAssertEqual(mockAPIService.passedAPIResponse, apiResponse)
    }
    
    func testAppCheckTokenResponseParsingError() async throws {
        let debugToken = UUID().uuidString
        let parsingError = NSError(domain: "testAppCheckTokenResponseParsingError", code: -1, userInfo: nil)
        
        let expectedRequestURL = "\(mockAPIService.baseURL)/projects/test_project_id/apps/test_app_id:exchangeDebugToken"
        let fakeResponseData = "fake response".data(using: .utf8)!
        let httpResponse = HTTPURLResponse(url: URL(string: expectedRequestURL)!, statusCode: 200, httpVersion: nil, headerFields: nil)!
        let apiResponse = _GACURLSessionDataResponse(response: httpResponse, httpBody: fakeResponseData)
        
        mockAPIService.sendRequestResult = .success(apiResponse)
        mockAPIService.appCheckTokenResult = .failure(parsingError)
        
        do {
            _ = try await debugAPIService.appCheckToken(debugToken: debugToken, limitedUse: false)
            XCTFail("Expected error to be thrown")
        } catch let error as NSError {
            XCTAssertEqual(error, parsingError)
        }
        
        XCTAssertEqual(mockAPIService.passedRequestURL?.absoluteString, expectedRequestURL)
        XCTAssertEqual(mockAPIService.passedHTTPMethod, "POST")
        XCTAssertEqual(mockAPIService.passedAdditionalHeaders?["Content-Type"], "application/json")
        try assertHTTPBody(mockAPIService.passedBody, debugToken: debugToken, limitedUse: false)
        XCTAssertEqual(mockAPIService.passedAPIResponse, apiResponse)
    }
    
    func testAppCheckTokenNetworkError() async throws {
        let debugToken = UUID().uuidString
        let networkError = NSError(domain: "testAppCheckTokenNetworkError", code: -1, userInfo: nil)
        
        mockAPIService.sendRequestResult = .failure(networkError)
        
        do {
            _ = try await debugAPIService.appCheckToken(debugToken: debugToken, limitedUse: false)
            XCTFail("Expected error to be thrown")
        } catch let error as NSError {
            XCTAssertEqual(error, networkError)
        }
        
        try assertHTTPBody(mockAPIService.passedBody, debugToken: debugToken, limitedUse: false)
    }
    
    // MARK: - Helpers
    
    func assertHTTPBody(_ body: Data?, debugToken: String, limitedUse: Bool) throws {
        let unwrappedBody = try XCTUnwrap(body)
        let decodedData = try JSONSerialization.jsonObject(with: unwrappedBody, options: []) as? [String: Any]
        let unwrappedDecodedData = try XCTUnwrap(decodedData)
        
        let decodeDebugToken = unwrappedDecodedData["debug_token"] as? String
        XCTAssertEqual(decodeDebugToken, debugToken)
        
        let decodedLimitedUse = unwrappedDecodedData["limited_use"] as? Bool
        XCTAssertEqual(decodedLimitedUse, limitedUse)
    }
}
